Update documentation for async api (#153)
oneapi-src/oneDPL: documentation/library_guide/extension_api.rst

        copy(oneapi::dpl::execution::dpcpp_default, view, range_res);
    }

Async API
--------------------------

Functions as defined in the STL algorithm or numeric headers are traditionally blocking. oneDPL extends the functionality of C++17 parallel algorithm by asynchronous algorithm with non-blocking behavior. This experimental feature enables expressing concurrent control flow by building dependency chains and interleaving algorithm calls, as well as interoperability with SYCL kernels. The following subset of C++17 STL algorithms are currently supported as async variants:

> * ``reduce``

Reviewer: I'd suggest to have alphabetical order for algorithms to seek an algorithm faster by user

Author: Yes makes sense. done.

* ``transform_reduce``
* ``sort``
* ``fill``
* ``copy``
* ``for_each``
* ``transform``

> The current Implementation for async algorithm is limited to device execution policy (DPCPP backend).

Reviewer: `device execution policy (DPCPP backend)` -> `DPC++ Execution Policies`

Author: done

* An arbitrary number of events can be passed to async algorithms as last arguments to allow expressing input dependencies.

The type of the object returned from asynchronous algorithm is undefined. Future-like, the following functions are present:

* ``get()``: returns the result
* ``wait()``: wait for the result to become available

If the returned object is the result of an algorithm with device policy, it can be converted into a ``sycl::event``. Lifetime of any resources the algorithm allocates (e.g. temporary storage) is bound to the lifetime of the returned object.

Utility functions:

* ``wait_for_all(…)``: wait for an arbitrary number of events to become ready.

Example of Async API usage
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: cpp

    #include <oneapi/dpl/execution>
    #include <oneapi/dpl/async>
    #include <CL/sycl.hpp>
    
    int main() {
        using namespace oneapi;
        {
            /* Build and compute a simple dependency chain: Fill buffer -> Transform -> Reduce */
            sycl::buffer<int> a{10};
            auto my_policy = dpl::execution::dpcpp_default;

>             auto fut1 = dpl::experimental::fill_async(my_policy,dpl::begin(a),dpl::end(a),7);

Reviewer: I'd suggest to mention explicitly in the documentation that async algorithms are available in `oneapi::dpl::experimental::` namespace

        copy(oneapi::dpl::execution::dpcpp_default, view, range_res);
    }

Async API
--------------------------

> Functions defined in the STL ``<algorithm>`` or ``<numeric>`` headers are traditionally blocking. oneDPL extends the functionality of C++17 parallel algorithms by providing asynchronous algorithm with non-blocking behavior. This experimental feature enables expressing concurrent control flow by building dependency chains and interleaving algorithm calls, as well as interoperability with SYCL kernels. 

Reviewer 2: This should be updated so it does not all run on one long line. Please add the following updates (in **bold**):

Functions defined in the STL ``<algorithm>`` or ``<numeric>`` headers are traditionally blocking. oneDPL extends the functionality of C++17 parallel algorithms by providing **an** asynchronous algorithm with non-blocking behavior. This experimental feature enables **you to express a** concurrent control flow by building dependency chains and interleaving algorithm calls **and** interoperability with SYCL***** kernels. [DB question: is it SYLC kernels or DPC++ kernels?]

Author: Both are supported: SYCL and DPC++ kernels. Also other interfaces from oneAPI that return a `sycl::event`, for example: oneMKL.

The current implementation for async algorithms is limited to DPC++ Execution Policies.

> All functionality described below is available in ``oneapi::dpl::experimental`` namespace.

Reviewer 2: Please add the bold below: 

All **the** functionality described below is available in **the** ``oneapi::dpl::experimental`` namespace.

The following async algorithms are currently supported:

* ``copy_async``
* ``fill_async``
* ``for_each_async``
* ``reduce_async``
* ``transform_async``
* ``transform_reduce_async``
* ``sort_async``

> All interfaces listed above are a subset of C++17 STL algorithms

Reviewer 2: Please add the bold below:

All **the** interfaces listed above are a subset of C++17 STL algorithms**,**

> with the suffix ``_async`` added to the corresponding name (e.g. ``reduce``, ``sort``, ... ). The behavior and signatures are overlapping with the C++17 STL algorithm with the following changes:

Reviewer 2: Please split this so it isn't a long line and add the bold below: 

**where** the suffix ``_async`` **is** added to the corresponding name (**for example:** ``reduce``, ``sort``, **etc.** ). The behavior and signatures overlap the C++17 STL algorithm with the following changes:

* Do not block the execution.
* Take an arbitrary number of events (including 0) as last arguments to allow expressing input dependencies.
* Return future-like object that allows ``wait`` for completion and ``get`` the result.

The type of the future-like object returned from asynchronous algorithm is unspecified. The following member functions are present:

* ``get()``: returns the result.
* ``wait()``: wait for the result to become available.

> If the returned object is the result of an algorithm with device policy, it can be converted into a ``sycl::event``. Lifetime of any resources the algorithm allocates (e.g. temporary storage) is bound to the lifetime of the returned object.

Reviewer 2: Please split this so it isn't a long line and add the bold below:

If the returned object is the result of an algorithm with device policy, it can be converted into a ``sycl::event``. Lifetime of any resources the algorithm allocates (**for example:** temporary storage) is bound to the lifetime of the returned object.

Utility functions:

* ``wait_for_all(…)``: wait for an arbitrary number of objects that are convertible into ``sycl::event`` to become ready.

> Example of Async API usage

Reviewer 2: Headers should be in title case: Example of Async API **U**sage

The following async algorithms are currently supported:

* ``copy_async``
* ``fill_async``
* ``for_each_async``
* ``reduce_async``
* ``transform_async``
* ``transform_reduce_async``
* ``sort_async``

All interfaces listed above are a subset of C++17 STL algorithms
with the suffix ``_async`` added to the corresponding name (e.g. ``reduce``, ``sort``, ... ). The behavior and signatures are overlapping with the C++17 STL algorithm with the following changes:

* Do not block the execution.
* Take an arbitrary number of events (including 0) as last arguments to allow expressing input dependencies.
* Return future-like object that allows ``wait`` for completion and ``get`` the result.

The type of the future-like object returned from asynchronous algorithm is unspecified. The following member functions are present:

* ``get()``: returns the result.
* ``wait()``: wait for the result to become available.

If the returned object is the result of an algorithm with device policy, it can be converted into a ``sycl::event``. Lifetime of any resources the algorithm allocates (e.g. temporary storage) is bound to the lifetime of the returned object.

Utility functions:

> * ``wait_for_all(…)``: wait for an arbitrary number of ``sycl::event`` to become ready.

Reviewer 3: I would like to make sure that this function cannot work with something but `sycl::event`.  Can we guarantee that it can work with something that is convertible to `sycl::event`? Otherwise, it's not so useful for our future-like objects.

Example of Async API usage
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: cpp

    #include <oneapi/dpl/execution>
    #include <oneapi/dpl/async>
    #include <CL/sycl.hpp>
    
    int main() {
        using namespace oneapi;
        {
            /* Build and compute a simple dependency chain: Fill buffer -> Transform -> Reduce */
            sycl::buffer<int> a{10};
            

>             auto my_policy = dpl::execution::dpcpp_default;

Reviewer 3: Do we need this line? I understand the idea behind that. You don't want to put `dpl::execution::dpcpp_default` in each algorithm call. I am not strongly suggesting direct passing without additional variable but probably it would look better. If you disagree fill free to resolve this conversation.

        copy(oneapi::dpl::execution::dpcpp_default, view, range_res);
    }

Async API
--------------------------

> Functions as defined in the STL ``<algorithm>`` or ``<numeric>`` headers are traditionally blocking. oneDPL extends the functionality of C++17 parallel algorithms by asynchronous algorithm with non-blocking behavior. This experimental feature enables expressing concurrent control flow by building dependency chains and interleaving algorithm calls, as well as interoperability with SYCL kernels. 

Reviewer 4: ```suggestion
Functions defined in the STL ``<algorithm>`` or ``<numeric>`` headers are traditionally blocking. oneDPL extends the functionality of C++17 parallel algorithms by providing asynchronous algorithm with non-blocking behavior. This experimental feature enables expressing concurrent control flow by building dependency chains and interleaving algorithm calls, as well as interoperability with SYCL kernels. 
```

> The current implementation for async algorithm is limited to DPC++ Execution Policies.

Reviewer 4: ```suggestion
The current implementation for async algorithms is limited to DPC++ Execution Policies.
```

All functionality described below is available in ``oneapi::dpl::experimental`` namespace.

The following async algorithms are currently supported:

* ``copy_async``
* ``fill_async``
* ``for_each_async``
* ``reduce_async``
* ``transform_async``
* ``transform_reduce_async``
* ``sort_async``

All interfaces listed above are a subset of C++17 STL algorithms

> with adding a suffix ``_async`` to the corresponding name (e.g. ``reduce``, ``sort``, ... ). The behavior and signatures are overlapping with the C++17 STL algorithm with the following changes:

Reviewer 4: ```suggestion
with the suffix ``_async`` added to the corresponding name (e.g. ``reduce``, ``sort``, ... ). The behavior and signatures are overlapping with the C++17 STL algorithm with the following changes:
```
